Pkcs7 coverage

nick-child-ibm opened the pull request:
Fixes #6745 .

Increase coverage of the pkcs7.c file:
|       | %lines |%branches|
| ----------- | ----------- | ----------- |
| before      | 85.7        | 68.5 |
| now   | 93.5         |  86.5 |

In this pull request, I have included a script which I have used to auto generate more tests in `test_suite_pkcs7.data`.
For every asn1 tag it adds tests to modify the length of the content and the tag type. When parsing this modified PKCS7 data buffer, it is expected to fail.

In some cases, the PKCS7 parsing functions did not pick up on the malformed asn1 data. I addressed these issues in the last commit of this pull request.

DemiMarie commented:
I wonder which branches are still not covered.

DemiMarie commented:
Could something similar be used for other places where Mbed TLS parses ASN.1?  I am specifically thinking of keys, algorithm identifiers, certificates, and CRLs.

daverodgman commented:
Please note, now that we have updated Mbed TLS with the new code style, there are conflicts. You can address these with the tool provided here: https://github.com/Mbed-TLS/mbedtls-docs/blob/main/tools/bin/mbedtls-rewrite-branch-style and then force-pushing

daverodgman commented on library/pkcs7.c:
```diff
@@ -541,13 +556,21 @@ int mbedtls_pkcs7_parse_der( mbedtls_pkcs7 *pkcs7, const unsigned char *buf,
         goto try_data;
     }
 
+    /* ensure bufferlength is only pkcs7 data */
```
Wording is a bit unclear. Suggest e.g.: "Ensure PKCS7 data uses the exact number of bytes specified in buflen"

daverodgman replied:
This requirement should be documented in pkcs7.h - otherwise users may think it's OK to pass a buffer with some empty space at the end.

daverodgman commented on tests/suites/test_suite_pkcs7.data:
```diff
@@ -64,11 +64,11 @@ pkcs7_parse:"data_files/pkcs7_signerInfo_serial_invalid_size.der":MBEDTLS_ERR_PK
 pkcs7_get_signers_info_set error handling (4541044530479104)
 depends_on:MBEDTLS_RIPEMD160_C
-pkcs7_parse:"data_files/pkcs7_get_signers_info_set-leak-fuzz_pkcs7-4541044530479104.der":MBEDTLS_ERR_PKCS7_INVALID_SIGNER_INFO
+pkcs7_parse:"data_files/pkcs7_get_signers_info_set-leak-fuzz_pkcs7-4541044530479104.der":MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO
```
The documentation for `MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO` in `pkcs7.h` could be improved: `The PKCS7 content info invalid or cannot be parsed.` - missing "is"?

daverodgman commented on tests/suites/test_suite_pkcs7.function:
```diff
@@ -26,6 +26,15 @@ int pkcs7_parse_buffer( unsigned char *pkcs7_buf, int buflen )
 /* END_SUITE_HELPERS */
 
+/* BEGIN_CASE */
+void pkcs7_asn1_fail( data_t *pkcs7_buf ) {
```
I would really like to see this testing for memory leaks, but unfortunately I'm not sure we have a good way to do this in our test code at the moment. I've manually verified that all of these tests perform the same number of `mbedtls_free` as `mbedtls_calloc`, i.e., do not appear to have an obvious memory leak.

nick-child-ibm replied:
I have been using `valgrind ./test_suite_pkcs7`, it has identified a couple of issues during development. No errors now though.

daverodgman commented on tests/scripts/generate_pkcs7_tests.py:
```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/env python3
```
How was the branch coverage info obtained?

daverodgman reviewed (changes requested):
Needs updating for code-style, plus some minor tweaks requested, and some questions

daverodgman commented:
The clean-up block in `pkcs7_get_signers_info_set()` is not well-covered (specifically, the while loop never executes). I'd like to see that one covered.

Otherwise the other gaps don't look too concerning - examples are things like a particular error code not being returned in a function that does return an error code which is tested elsewhere (so the fact that the function can return an error is tested) - e.g. `return MBEDTLS_ERR_PKCS7_INVALID_ALG` in `pkcs7_get_signed_data()`. So these don't look like a must-fix to me.

nick-child-ibm commented:
Thanks for the review! I believe I have addressed all of the points made, please correct me if wrong.
Coverage is up to `94.7 %` of lines and `87.1%` of branches.

If requested, I can squash some commits and add the final coverage deltas.

@DemiMarie I am not sure if using the ASN1 python script in other areas is such a good idea.  The script itself is a bit specific to PKCS7 at the moment, I am sure it can fixed up to extend the scope to other ASN1 structures but even then the script changes tags and lengths and not much else. There are probably some corner cases where changing the values could still result in legal structures. It is a nice thing in concept but it could get a bit sticky (and make your test files a lot grosser).

DemiMarie commented:
@nick-child-ibm: what would it take to reach 100% branch coverage?

DemiMarie commented on library/pkcs7.c:
```diff
@@ -312,6 +313,7 @@ static int pkcs7_get_signer_info(unsigned char **p, unsigned char *end,
+    end_issuer_and_sn = *p + len;
     /* Parsing IssuerAndSerialNumber */
     signer->issuer_raw.p = *p;
 
```
`end_signer` should be replaced by `end_issuer_and_sn` in various places.  This is harmless inasmuch as any mismatches will be caught by the `*p != end_issuer_and_sn` check.

DemiMarie commented:
Also, is there a check that the signature has no inline data?

gilles-peskine-arm commented:
> what would it take to reach 100% branch coverage?

We don't have a way to make memory allocations fail on demand, so error paths that can only be triggered by a memory allocation failure are excused from test coverage for the time being.

tom-cosgrove-arm commented:
Need to fix the pylint complaint

```
tests/scripts/generate_pkcs7_tests.py:41: error: Format strings are only supported in Python 3.6 and greater
Found 1 error in 1 file (checked 29 source files)
^^^^check_python_files: Lint: Python scripts: tests/scripts/check-python-files.sh -> 1^^^^
```

daverodgman commented on library/pkcs7.c:
```diff
@@ -574,6 +594,12 @@ int mbedtls_pkcs7_parse_der(mbedtls_pkcs7 *pkcs7, const unsigned char *buf,
+        ret = MBEDTLS_ERR_PKCS7_BAD_INPUT_DATA;
+        goto out;
+    }
+
```
At line 663 / 668, we have untested error paths for an unrecognised digest algorithm. It looks like these should be straight-forward to add test-cases for by adding PCKS7 files using disabled and invalid digest algs and using the existing pkcs7_parse test case.

nick-child-ibm replied:
Agree would be nice to test this. Unfortunately, this block is in the verify function, so we would have to first have it pass the parsing function to get the pkcs7 struct. In parsing logic, unsupported/invalid digest algs are being caught by `pkcs7_get_signed_data`:
```
ret = pkcs7_get_digest_algorithm_set(&p, end_set,
                                         &signed_data->digest_alg_identifiers);
    if (ret != 0) {
        return ret;
    }

    ret = mbedtls_oid_get_md_alg(&signed_data->digest_alg_identifiers, &md_alg);
    if (ret != 0) {
        return MBEDTLS_ERR_PKCS7_INVALID_ALG;
    }
```
But   ` return MBEDTLS_ERR_PKCS7_INVALID_ALG` was not getting covered here either. So I will add a test for this.

DemiMarie replied:
If this is genuinely unreachable, should there be an assertion?

daverodgman replied:
It could be reached by manually modifying fields in the pkcs7 structure I guess (indeed, this would be the way to test it). And in general, Mbed TLS does not use assertions. So I think the approach is correct.

nick-child-ibm commented:
> Also, is there a check that the signature has no inline data?

@DemiMarie 
Not sure what specific case is in question, with ASN1 there are a couple invalid structures to think about:
If the lengths of the various ASN1 tags which encompass the signature imply that a signature is present when there is no signature data (but a ASN1 tag for the signature is present ) then conditionals like `if (*p != end_signer)` will trigger an error.

Or, if the lengths of the various ASN1 tags which encompass the signature imply that there is no signature present AND the signature tag is nonexistent then we will error in `pkcs7_get_signature` when we fetch tag `MBEDTLS_ASN1_OCTET_STRING`.

On the other hand, if the various ASN1 tags which encompass the signature imply that there is no signature present AND the tag is correct with the data being empty then.. well that is interesting to think about. I had to go and manually test this to see what would happen. It will parse successfully and fail to verify which is good (also tried on openssl just for fun, same behavior). So maybe we should discuss whether MbedTLS should error out in these circumstances.

Sorry I hope I answered your question.

DemiMarie commented:
@nick-child-ibm PKCS#7 signatures can either contain the data that is signed, or merely contain a hash of that data.  Mbed TLS only supports the second case, so it should reject the first.

daverodgman commented on library/pkcs7.c:
```diff
@@ -546,12 +559,19 @@ int mbedtls_pkcs7_parse_der(mbedtls_pkcs7 *pkcs7, const unsigned char *buf,
+    }
+
     if (!MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_DATA, &pkcs7->content_type_oid)
         || !MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_ENCRYPTED_DATA, &pkcs7->content_type_oid)
```
Just spotted, we have `MBEDTLS_OID_PKCS7_ENCRYPTED_DATA` twice in this list. This is harmless, but please correct it. Wondering if it would be clearer to invert the whole condition block and write it as:

```
if (MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_SIGNED_DATA, &pkcs7->content_type_oid)) {

DemiMarie replied:
It indeed would be.

nick-child-ibm replied:
I will remove the duplicate entry. But the conditional is important:
```
if (!MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_DATA, &pkcs7->content_type_oid)
        || !MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_ENVELOPED_DATA, &pkcs7->content_type_oid)
        || !MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_SIGNED_AND_ENVELOPED_DATA, &pkcs7->content_type_oid)
        || !MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_DIGESTED_DATA, &pkcs7->content_type_oid)
        || !MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_ENCRYPTED_DATA, &pkcs7->content_type_oid))
```
returns `MBEDTLS_ERR_PKCS7_FEATURE_UNAVAILABLE` because these are legal oid's but the parsing logic is not implemented.
At the moment, the next conditional is:
```
if (MBEDTLS_OID_CMP(MBEDTLS_OID_PKCS7_SIGNED_DATA, &pkcs7->content_type_oid))
```
Which returns `MBEDTLS_ERR_PKCS7_BAD_INPUT_DATA`, because the oid is not a legal PKCS7 type.

DemiMarie replied:
The fast path is `MBEDTLS_OID_PKCS7_SIGNED_DATA`, so that should be the outer conditional.

daverodgman replied:
True, but I don't think saving a handful of instructions at the start of parsing a PKCS7 file will make a noticeable difference so I'm happy with how it's done in the latest commit.

daverodgman commented:
> PKCS#7 signatures can either contain the data that is signed, or merely contain a hash of that data. Mbed TLS only supports the second case, so it should reject the first.

Good point - I've checked this and it does reject (with `MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO`) - but there is no test case for this. Agree that we ought to add this please.

DemiMarie commented:
> > PKCS#7 signatures can either contain the data that is signed, or merely contain a hash of that data. Mbed TLS only supports the second case, so it should reject the first.
> 
> Good point - I've checked this and it does reject (with `MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO`) - but there is no test case for this. Agree that we ought to add this please (should simply be a case of adding a suitable der file, and use the existing `pkcs7_parse` test case)

`MBEDTLS_ERR_PKCS7_FEATURE_UNAVAILABLE` would be a more useful error: the signature is syntactically correct, but Mbed TLS cannot handle it.

nick-child-ibm commented:
> > > supports the second case, so it should reject the first.
> > 
> > 
> > Good point - I've checked this and it does reject (with `MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO`) - but there is no test case for this. Agree that we ought to add this please (should simply be a case of adding a suitable der file, and use the existing `pkcs7_parse` test case)
> 
> `MBEDTLS_ERR_PKCS7_FEATURE_UNAVAILABLE` would be a more useful error: the signature is syntactically correct, but Mbed TLS cannot handle it.

Sorry a bit lost here. If the signature was just a hash then the fields specifying the signature algs would be whatever the OID for NONE is. Is that correct? Was this tested? In this case, I would expect this to be handled without error. That being said, I don't know why anyone would want to do this or what the purpose would be.

DemiMarie commented:
> > > > supports the second case, so it should reject the first.
> > > 
> > > 
> > > Good point - I've checked this and it does reject (with `MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO`) - but there is no test case for this. Agree that we ought to add this please (should simply be a case of adding a suitable der file, and use the existing `pkcs7_parse` test case)
> > 
> > 
> > `MBEDTLS_ERR_PKCS7_FEATURE_UNAVAILABLE` would be a more useful error: the signature is syntactically correct, but Mbed TLS cannot handle it.
> 
> Sorry a bit lost here. If the signature was just a hash then the fields specifying the signature algs would be whatever the OID for NONE is. Is that correct?

NONE is not a valid signature algorithm, as it provides no security.  Therefore, it must be rejected.

> Was this tested? In this case, I would expect this to be handled without error. That being said, I don't know why anyone would want to do this or what the purpose would be.

I am referring to a PKCS#7 signature with inline data.  Mbed TLS does not support this, so `MBEDTLS_ERR_PKCS7_FEATURE_UNAVAILABLE` is a better error than `MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO`.  This should only impact the error path.

daverodgman commented:
> > > > > supports the second case, so it should reject the first.
> > > > 
> > > > 
> > > > Good point - I've checked this and it does reject (with `MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO`) - but there is no test case for this. Agree that we ought to add this please (should simply be a case of adding a suitable der file, and use the existing `pkcs7_parse` test case)
> > > 
> > > 
> > > `MBEDTLS_ERR_PKCS7_FEATURE_UNAVAILABLE` would be a more useful error: the signature is syntactically correct, but Mbed TLS cannot handle it.
> > 
> > 
> > Sorry a bit lost here. If the signature was just a hash then the fields specifying the signature algs would be whatever the OID for NONE is. Is that correct?
> 
> NONE is not a valid signature algorithm, as it provides no security. Therefore, it must be rejected.
> 
> > Was this tested? In this case, I would expect this to be handled without error. That being said, I don't know why anyone would want to do this or what the purpose would be.
> 
> I am referring to a PKCS#7 signature with inline data. Mbed TLS does not support this, so `MBEDTLS_ERR_PKCS7_FEATURE_UNAVAILABLE` is a better error than `MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO`. This should only impact the error path.

FYI, I was able to construct a test case with:

`openssl smime -sign -md sha256 -binary -nocerts -noattr -in test_data -out pkcs7_data_with_signature.der -outform der -inkey pkcs7-rsa-sha256-1.key -signer pkcs7-rsa-sha256-1.crt -nodetach`

and then:

```
PKCS7 Signed non-detached
depends_on:MBEDTLS_SHA256_C:MBEDTLS_RSA_C
pkcs7_parse:"data_files/pkcs7_data_with_signature.der":MBEDTLS_ERR_PKCS7_FEATURE_UNAVAILABLE
```

nick-child-ibm commented:
Ahh okay, I thought you were referring to the `EncryptedDigest` field of `signerInfos`. I see now it is the `contentInfo` field of `signedData` which can optionally contain additional data. I made a note of this limitation in the header file and fixed the error value.

I also rebased (sorry) because I missed some f-strings in the "Remove f strings" commit.

Let me know if an additional test should be added for an invalid digest algorithm. I added a test for unsupported digest algorithm and figured it was essentially the same thing from the execution point of view.

nick-child-ibm commented:
Are the python script formatting errors worth fixing? Is it getting merged?

tom-cosgrove-arm commented:
> Are the python script formatting errors worth fixing? Is it getting merged?

Yes please - we need the CI to pass before we can merge, and it won't while `check-python-files.sh` continues to complain

daverodgman commented on library/pkcs7.c:
```diff
@@ -481,11 +489,16 @@ static int pkcs7_get_signed_data(unsigned char *buf, size_t buflen,
+    ret = pkcs7_get_content_info_type(&p, end_set, &end_content_info,
+                                      &signed_data->content.oid);
     if (ret != 0) {
         return ret;
```
I think the issue is that here, we need to distinguish between valid content present (which is unsupported), vs invalid content (invalid data). Something like the following would work:

```suggestion
        /* Determine if valid content is present */
        ret = mbedtls_asn1_get_tag(&p, end_content_info, &len, MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_CONTEXT_SPECIFIC);
        if (ret != 0) {
            return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO, ret);
        }
        ret = mbedtls_asn1_get_tag(&p, end_content_info, &len, MBEDTLS_ASN1_OCTET_STRING);
        if (ret != 0) {
            return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO, ret);
        }
        p += len;
        if (p != end_content_info) {
            return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO, ret);
        }
        /* Valid content is present - this is not supported */
        return MBEDTLS_ERR_PKCS7_FEATURE_UNAVAILABLE;
```

(this would mean reverting the changes to the expected result for the two `pkcs7_get_signers_info_set error handling` cases, which I think makes sense).

The only downside is that this introduces a couple of uncovered lines - but as they are returning errors in a block that can only return an error, I'm OK with this.

daverodgman replied:
@tom-cosgrove-arm WDYT?

DemiMarie replied:
> The only downside is that this introduces a couple of uncovered lines - but as they are returning errors in a block that can only return an error, I'm OK with this.

I would still prefer to see tests for them, especially because future changes may add support for embedded data.

tom-cosgrove-arm replied:
Actually, in this case, let's get the code in and work on those tests later, please

daverodgman reviewed (dismissed):
Assuming CI passes, LGTM

daverodgman commented:
@nick-child-ibm sorry, looks like there are still a few python formatting issues

daverodgman reviewed (approved):
Updates look good, assuming CI is happy

tom-cosgrove-arm reviewed (approved):
LGTM, contingent on CI

daverodgman commented:
Looks like one pylint issue is still remaining:

```
Running pylint ...
************* Module generate_pkcs7_tests
tests/scripts/generate_pkcs7_tests.py:60:4: R0201: Method could be a function (no-self-use)

------------------------------------
Your code has been rated at 10.00/10

pylint reported errors
```

daverodgman commented on tests/scripts/generate_pkcs7_tests.py:
```diff
@@ -0,0 +1,194 @@
+        self.last_test_num, self.old_tests = self.read_test_file(file_name)
+        self.new_tests = []
+
+    def read_test_file(self, file):
```
```suggestion
    # pylint: disable=no-self-use
    def read_test_file(self, file):
```

tom-cosgrove-arm reviewed (approved):
LGTM

nick-child-ibm commented:
One of these days I will actually take the time to install the right dependencies to properly run the tests locally. Apologies and thanks for fixing. On my end it was not reporting any issues, I suspect I have a version mismatch somewhere.
I hope to one day to have as almighty of a CI in my projects.
 
As always, thanks again for the wonderful reviews.

tom-cosgrove-arm reviewed (approved):
LGTM
